autobahn/wamp/protocol.py

```diff
@@ -490,6 +490,32 @@ def _swallow_error(self, fail, msg):
             )
         return None
 
+    def type_check(self, func, return_error=False):
+        """
+        Does parameter type checking and validation against type hints
+        and appropriately tells the user code and the caller (through router).
+
+        :return_error: Whether the error should be returned to the caller in
+                       addition to being notified to the user code.
+        :type: bool
+        """
+        async def _type_check(*args, **kwargs):
+            # Converge both args and kwargs into a dictionary
+            arguments = inspect.getcallargs(func, *args, **kwargs)
+            response = []
+            for name, kind in func.__annotations__.items():
+                if name in arguments and not isinstance(arguments[name], kind):
+                    response.append(
+                        "'{}' required={} got={}".format(name, kind.__name__, type(arguments[name]).__name__))
+            if response:
+                error = TypeCheckError(', '.join(response))
+                self.onUserError(txaio.create_future_error(error))
+                if return_error:
```

Comment:
I haven't completely thought through this yet, but don't we just want to `raise TypeCheckError()` here, and let the framework deal with the error the way it normally does? That would imply getting rid of `return_error` too since then it would always  be `True`. (Maybe I'm just asking: why the distinction here?)

Reply:
I had changed that per this comment https://github.com/crossbario/autobahn-python/pull/1307#discussion_r375840597

Reply:
@oberstet ^^

Reply:
yes, just raise TypeCheckError and let existing framework code handle the rest - so that we don't duplicate behavior in the type-check decorator

Reply:
Fixed that with a following commit https://github.com/crossbario/autobahn-python/pull/1314